pylint/config/utils.py

```diff
@@ -0,0 +1,75 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+
+import sys
+from typing import TYPE_CHECKING, List, Optional, Union
+
+from pylint import reporters
+from pylint.utils import utils
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+
+def _config_initialization(
+    linter: "PyLinter",
+    args_list: List[str],
+    reporter: Union[reporters.BaseReporter, reporters.MultiReporter, None] = None,
+    config_file: Optional[str] = None,
+    verbose_mode: Optional[bool] = None,
+) -> List[str]:
+    """Parse all available options, read config files and command line arguments and
+    set options accordingly.
+    """
+
+    # Read the config file. The parser is stored on linter.cfgfile_parser
+    try:
+        linter.read_config_file(config_file=config_file, verbose=verbose_mode)
+    except OSError as ex:
+        print(ex, file=sys.stderr)
+        sys.exit(32)
+    config_parser = linter.cfgfile_parser
+
+    # Run init hook, if present, before loading plugins
+    if config_parser.has_option("MASTER", "init-hook"):
+        exec(  # pylint: disable=exec-used
+            utils._unquote(config_parser.get("MASTER", "init-hook"))
```

Comment:
The init-hook is not tested as it's just been added, maybe we could add a small functional test for it ?

Reply:
Got any good ideas? Not really sure what people use it for anyway.

Reply:
Btw, it needs to be a functional test with an `init-hook`. So just `raise RuntimError` like we do in another test won't work sadly...

Reply:
I know that people used it to modify the sys path, in order to increase max inference in astroid, the maximum recursion depth for genuinely large recursion trees... Well that's the one I hear about in issues anyway (all in order to avoid problem with pylint as you see). There probably others.

Reply:
I found a way to do it :)